src/bms/ast/structure.rs

```diff
@@ -0,0 +1,75 @@
+//! The structure of the AST.
+
+use std::collections::BTreeMap;
+
+use num::BigUint;
+
+use crate::bms::{command::mixin::SourcePosMixin, lex::token::TokenWithPos};
+
+/// An unit of AST which represents individual scoped commands of BMS source.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Unit<'a> {
+    /// A token that is not a control flow token.
+    TokenWithPos(&'a TokenWithPos<'a>),
+    /// A Random block. Can contain multiple If blocks.
+    RandomBlock {
+        /// The value of the Random block.
+        value: SourcePosMixin<BlockValue>,
+        /// The If blocks in the Random block.
+        if_blocks: Vec<IfBlock<'a>>,
+    },
+    /// A Switch block.
+    /// Like C++ Programming Language, Switch block can contain multiple Case branches, and a Def branch.
+    /// If there is no other Case branch activated, Def branch will be activated.
+    /// When executing, the tokens, from the activated branch, to Skip/EndSwitch, will be executed.
+    SwitchBlock {
+        /// The value of the Switch block.
+        value: SourcePosMixin<BlockValue>,
+        /// The Case branches in the Switch block.
+        cases: Vec<CaseBranch<'a>>,
```

Comment:
To preserve source position information, you may need these fields:
```suggestion
        /// The ENDIF position in the BMS source if exists.
        end_if: Option<SourcePosMixin<()>>,
    },
    /// A Switch block.
    /// Like C++ Programming Language, Switch block can contain multiple Case branches, and a Def branch.
    /// If there is no other Case branch activated, Def branch will be activated.
    /// When executing, the tokens, from the activated branch, to Skip/EndSwitch, will be executed.
    SwitchBlock {
        /// The value of the Switch block.
        value: SourcePosMixin<BlockValue>,
        /// The Case branches in the Switch block.
        cases: Vec<CaseBranch<'a>>,
        /// The ENDSW position in the BMS source.
        end_sw: SourcePosMixin<()>,
```